---
title: "RNAseq Pipeline Report"
author: "Michal Varga"
date: "`r Sys.Date()`"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(kableExtra)
library(rmarkdown)
```

# Introduction

# Preprocessing

do this last

# Differential Expression Analysis

## Load Required Libraries

Install scientific packages via `BiocManager`.

```{r libraries_biocManager_install, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq2")
BiocManager::install("org.Bt.eg.db")
BiocManager::install("vsn")
```

Load `BiocManager` related libraries.

```{r libraries_biocManager_load, message=FALSE}
library(DESeq2)
library(org.Bt.eg.db)
library(vsn)
```

Load `CRAN` related libraries.

```{r libraries_other_load, message=FALSE}
library(here)
library(RColorBrewer)
library(tidyverse)
library(pheatmap)
library(data.table)
```

Load custom scripts and define colour palette.

```{r volcano_and_palette, message=FALSE}
here::i_am("README.md")
source(here("scripts/volcano_plot.R"))
pal <- brewer.pal(11,"RdBu")
```


## Load Count Data and Metadata

1. Load count data and change the order of conditions.

```{r load_counts}
data <- as.matrix(read.table('../Data/WT40vsMock.counts', header = T))
data <- data[,sort(colnames(data))] # reorder the conditions so they go control, condition
```
```{r load_counts_display, echo=FALSE, warning=FALSE}
data |> head(10) |> kable()
```

2. Load count metadata and 

```{r load_metadata}
data.meta <- as.data.frame(read.csv('../Data/WT40_meta.csv', header = T))
data.meta <- data.meta |> arrange(Names) # reorder the conditions so they go control, condition
data.meta <- data.meta %>% remove_rownames %>% column_to_rownames(var="Names") # removes the colum name of sample names
```
```{r load_metadata_display, echo=FALSE, warning=FALSE}
data.meta |> kable()
```

filters out viral transcripts (only leaves stuff starting with ENS)
```{r}
nrow(data)
data <- data[rownames(data) %like% "ENS", ]
nrow(data)
```

check if count sample names and metdata sample names are the same
```{r}
all(rownames(data.meta) %in% colnames(data))
```

## Create DESeq2 Object

1. Create DESeq2 object.

```{r deseq2_object_creation}
dds <- DESeqDataSetFromMatrix(countData = data,
                              colData = data.meta,
                              design = ~ Condition)
```

2. Set `untreated` copndition as reference.

```{r dds_reference}
dds$Condition <- relevel(dds$Condition, ref = "untreated")
```

3. Filter genes with no counts.

```{r dds_filter_no_counts}
nrow(dds)
keep <- rowSums(counts(dds)) > 1
dds <- dds[keep,]
nrow(dds)
```

## Exploratory Analysis and Visualisation
### Data Transformation

Tools for analysis of multidimensional data require the same range of variance at different ranges of the mean values (`homoskedasticity`). Raw count data freom RNAseq expreriments, however, have variance that grows with the mean. Transforming the dataset with transformations like **Variance Stabilising Transformation** (VST) will lead in `homoskedastic` data, by poroducing log2-like values high counts.

#### Raw Count Data Skedasticity

```{r raw_dds_viz}
meanSdPlot(log(assay(dds)[rowSums(assay(dds))>30,]))
```

#### Variance Stabilising Transformation (VST) Processed Data Skedasticity

```{r vst_viz}
vsd <- DESeq2::vst(dds)
meanSdPlot(log(assay(vsd)[rowSums(assay(vsd))>0,]))
```

### Principal Component Analysis (PCA)

```{r pca}
DESeq2::plotPCA(vsd, intgroup = 'Condition')
```

We see that 92% of the variance is due to the principal component (PC) 1, which seems to be the infection state.

### Sample Distances

```{r sample_distances_creation}
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
colnames(sampleDistMatrix) <- NULL
```

```{r sample_distances_heatmap}
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

We again observe the infected and uninfected samples being similar within their conditions.

## Statistical Testing for Differences Attributable to Changes Between Conditions

```{r deseq2_differential_expression_object}
dds <- DESeq2::DESeq(dds)
```

### Dispersion Estimation

The assumption that most genes are not differentially expressed is assessed by plotting the dispersion estimation.


```{r dispersion_estimation}
DESeq2::plotDispEsts(dds)
```

### Results Retrieval

Following [Schurch et al., (RNA, 2016)](https://rnajournal.cshlp.org/content/22/6/839.full.pdf) recommendations, for 3 replicates per conditions:

```{r dds_results}
res <- results(dds)
res <- res[order(res$padj),]
head(res)
summary(res)
```

```{r dds_results_schurch}
resSchurch <- results(dds, lfcThreshold = 0.5, alpha = 0.01)
resSchurch <- resSchurch[order(resSchurch$padj),]
head(resSchurch)
summary(resSchurch)
```

```{r}
hist(resSchurch$pvalue,breaks=seq(0,1,.01))
```

### Visualisations

Visualisations assess the assumption that the majority of genes are not differentially expressed.

#### Volcano plot

```{r volcano_plot}
volcanoPlot(resSchurch)
```

#### MA plot

```{r ma_plot}
DESeq2::plotMA(resSchurch, ylim = c(-5, 5))
```

### Adding Gene Annotations


```{r ensemb_string}
ens.str <- substr(rownames(resSchurch),1,18)
```


```{r annot_symbol, message=FALSE}
resSchurch$symbol <- mapIds(org.Bt.eg.db,
                            keys=ens.str,
                            column="SYMBOL",
                            keytype="ENSEMBL",
                            multiVals="first")
```

```{r annot_genename, message=FALSE}
resSchurch$gene_name <- mapIds(org.Bt.eg.db,
                               keys=ens.str,
                               column="GENENAME",
                               keytype="ENSEMBL",
                               multiVals="first")
```

```{r annot_go, message=FALSE}
resSchurch$go <- mapIds(org.Bt.eg.db,
                        keys=ens.str,
                        column="GO",
                        keytype="ENSEMBL",
                        multiVals="first")
```

```{r annot_ontology, message=FALSE}
resSchurch$ontology <- mapIds(org.Bt.eg.db,
                              keys=ens.str,
                              column="ONTOLOGY",
                              keytype="ENSEMBL",
                              multiVals="first")
```


```{r annotated_res, echo=FALSE}
head(resSchurch, 5) |> kable()
```

```{r top_log2_change}
# subset significant results for resSchurch
resSigSchurch <- subset(resSchurch, padj < 0.01)
head(resSigSchurch[ order(resSigSchurch$log2FoldChange, decreasing = TRUE), ])
```

### Heatmap of Top Variable Genes

```{r top_var_heatmap}
topVarGenes <- head(order(rowVars(assay(vsd)), decreasing = TRUE), 15)

var_matrix  <- assay(vsd)[ topVarGenes, ]
var_matrix  <- var_matrix - rowMeans(var_matrix)

pheatmap(var_matrix)
```
