---
title: "Differential Expression Analysis of RNAseq Data"
author: "Michal Varga"
date: "`r Sys.Date()`"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(kableExtra)
library(rmarkdown)
```

# Introduction

This project presents a comprehensive differential expression analysis of RNA-seq data, a powerful approach for uncovering gene expression changes across different biological conditions. Using DESeq2, this analysis identifies genes with significant expression differences, shedding light on underlying biological processes, disease mechanisms, and treatment effects. By leveraging RNA-seq technology, we gain high-resolution insights into gene activity, enabling robust statistical testing and meaningful interpretation of complex biological datasets.

## Load Required Libraries and Scripts

0. (Optional) Install scientific packages via `BiocManager`.

```{r libraries_biocManager_install, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq2")
BiocManager::install("org.Bt.eg.db")
BiocManager::install("vsn")
```

1. Load `BiocManager` related libraries.

```{r libraries_biocManager_load, message=FALSE}
library(DESeq2)
library(org.Bt.eg.db)
library(vsn)
```

2. Load `CRAN` related libraries.

```{r libraries_other_load, message=FALSE}
library(here)
library(RColorBrewer)
library(tidyverse)
library(pheatmap)
library(data.table)
```

3. Load custom scripts and define colour palette.

```{r volcano_and_palette, message=FALSE}
here::i_am("README.md")
source(here("scripts/volcano_plot.R"))
pal <- brewer.pal(11,"RdBu")
```

## Load Count Data and Metadata

1. Load count data and change the order of conditions.

```{r load_counts}
data <- as.matrix(read.table('../Data/WT40vsMock.counts', header = T))
data <- data[,sort(colnames(data))] # reorder the conditions so they go control, condition
```
```{r load_counts_display, echo=FALSE, warning=FALSE}
data |> head(10) |> kable()
```

2. Load count metadata, arrange it as before, and remove `Names` column name. 
```{r load_metadata}
data.meta <- as.data.frame(read.csv('../Data/WT40_meta.csv', header = T))

# reorder the conditions so they go control, condition
data.meta <- data.meta |> arrange(Names)

# removes the column name of sample names
data.meta <- data.meta %>% remove_rownames %>% column_to_rownames(var="Names")
```
```{r load_metadata_display, echo=FALSE, warning=FALSE}
data.meta |> kable()
```

3. Filters out viral transcripts (only leaves genes starting with ENS).

```{r}
nrow(data)
data <- data[rownames(data) %like% "ENS", ]
nrow(data)
```

4. Assess if count sample names and metadata sample names are equal.

```{r}
all(rownames(data.meta) %in% colnames(data))
```

## Create DESeq2 Object

1. Create DESeq2 object.

```{r deseq2_object_creation}
dds <- DESeqDataSetFromMatrix(countData = data,
                              colData = data.meta,
                              design = ~ Condition)
```

2. Set `untreated` condition as reference.

```{r dds_reference}
dds$Condition <- relevel(dds$Condition, ref = "untreated")
```

3. Filter genes with no counts.

```{r dds_filter_no_counts}
nrow(dds)
keep <- rowSums(counts(dds)) > 1
dds <- dds[keep,]
nrow(dds)
```

# Exploratory Analysis and Visualisation
## Data Transformation

Tools for analysis of multidimensional data require the same range of variance at different ranges of the mean values (`homoskedasticity`). Raw count data freom RNAseq expreriments, however, have variance that grows with the mean. Transforming the dataset with transformations like **Variance Stabilising Transformation** (VST) will lead in `homoskedastic` data, by poroducing log2-like values high counts.

### Raw Count Data Skedasticity

```{r raw_dds_viz}
meanSdPlot(log(assay(dds)[rowSums(assay(dds))>30,]))
```

### Variance Stabilising Transformation (VST) Processed Data Skedasticity

```{r vst_viz}
vsd <- DESeq2::vst(dds)
meanSdPlot(log(assay(vsd)[rowSums(assay(vsd))>0,]))
```

## Principal Component Analysis (PCA)

```{r pca}
DESeq2::plotPCA(vsd, intgroup = 'Condition')
```

We see that 92% of the variance is due to the principal component (PC) 1, which seems to be the infection state.

## Sample Distances Plot

1. Prepare Sample Distances Matrix
```{r sample_distances_creation}
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
colnames(sampleDistMatrix) <- NULL
```

2. Plot it as heatmap.
```{r sample_distances_heatmap}
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

We again observe the infected and uninfected samples being similar within their conditions.

# Statistical Testing for Differences Attributable to Changes Between Conditions

1. Create DESeq2 differential expression object.
```{r deseq2_differential_expression_object}
dds <- DESeq2::DESeq(dds)
```

## Dispersion Estimation

In RNA-seq analysis, dispersion estimation is a critical step for accurately modeling gene expression variability. DESeq2 assumes that most genes are not differentially expressed, allowing it to estimate dispersion (or variability) more precisely by pooling information across genes.

Dispersion reflects the extent to which gene counts deviate from the expected mean expression level across replicates. In DESeq2, this is modeled using the Negative Binomial distribution, which accounts for overdispersion commonly observed in RNA-seq data. Accurate dispersion estimation enhances the detection of true differences in gene expression between conditions by stabilizing variability in low-count genes, reducing noise, and moderating extreme dispersion values.

The following plot shows gene-wise dispersion estimates against the mean expression level. The black points represent per-gene estimates, while the red line shows the fitted dispersion trend used by DESeq2 for normalization.

```{r dispersion_estimation}
DESeq2::plotDispEsts(dds)
```

## Results Retrieval

Following [Schurch et al., (RNA, 2016)](https://rnajournal.cshlp.org/content/22/6/839.full.pdf) recommendations, for 3 replicates per conditions we selected only genes with a log fold change (LFC) of more than 0.5, and adjusted p-value (alpha) of more than 0.01.

This way we have extracted the list of statistically significant, differentialy expressed genes between the treated and untreated conditions.

1. Create results object with the thresholds mentioned above and order the genes based on ascending adjusted p-value.
```{r dds_results_schurch}
resSchurch <- results(dds, lfcThreshold = 0.5, alpha = 0.01)
resSchurch <- resSchurch[order(resSchurch$padj),]
```

2. Evaluate the obtained results.
```{r results_schurch_summary}
head(resSchurch)
summary(resSchurch)
```

3. Construct histogram of p-values to assess their distribution.
```{r histogram_pvals}
hist(resSchurch$pvalue,breaks=seq(0,1,.01))
```

## Visual Assesment of Differentialy Expressed genes

Visualisations assess the assumption that the majority of genes are not differentially expressed.

### Volcano plot

```{r volcano_plot}
volcanoPlot(resSchurch)
```

### MA plot

```{r ma_plot}
DESeq2::plotMA(resSchurch, ylim = c(-5, 5))
```

## Adding Gene Annotations
To obtain gene annotation data that in useful for possible downstream gene enrichment analyses, we called `mapIds` function.

1. Create a vector of ENSEMBL IDs from results object.
```{r ensemb_string}
ens.str <- substr(rownames(resSchurch),1,18)
```

2. Obtain gene symbol annotation.
```{r annot_symbol, message=FALSE}
resSchurch$symbol <- mapIds(org.Bt.eg.db,
                            keys=ens.str,
                            column="SYMBOL",
                            keytype="ENSEMBL",
                            multiVals="first")
```

3. Obtain gene name annotation.
```{r annot_genename, message=FALSE}
resSchurch$gene_name <- mapIds(org.Bt.eg.db,
                               keys=ens.str,
                               column="GENENAME",
                               keytype="ENSEMBL",
                               multiVals="first")
```

4. Obtain gene ontology ID annotation.
```{r annot_go, message=FALSE}
resSchurch$go <- mapIds(org.Bt.eg.db,
                        keys=ens.str,
                        column="GO",
                        keytype="ENSEMBL",
                        multiVals="first")
```

5. Obtain gene ontology type annotation.
```{r annot_ontology, message=FALSE}
resSchurch$ontology <- mapIds(org.Bt.eg.db,
                              keys=ens.str,
                              column="ONTOLOGY",
                              keytype="ENSEMBL",
                              multiVals="first")
```
```{r annotated_res, echo=FALSE}
head(resSchurch, 5) |> kable()
```

## Heatmap of Top Variable Genes

```{r top_var_heatmap}
topVarGenes <- head(order(rowVars(assay(vsd)), decreasing = TRUE), 15)

var_matrix  <- assay(vsd)[ topVarGenes, ]
var_matrix  <- var_matrix - rowMeans(var_matrix)

pheatmap(var_matrix)
```
